twister: DeviceHandler DUT selection improvements and fixes

golowanow opened the pull request:
Several improvements and fixes for how Twister DeviceHandler selects and releases a DUT having multiple candidates for a test instance execution and taking into account failures.
See individual commits for details.

golowanow committed:
twister: Improve DUT release at DeviceHandler

Several improvements at Twister DeviceHandler when it releases
current DUT (Device Under Test):

 - release the exact DUT which is used for the test instance
   instead of all configured DUTs which happened to have
   the same serial device configured.

 - Twister PyTtest harness plugin adjustment to the above.

 - additional debug logging to track DUT waiting/retain/release.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
twister: Fix DeviceHandler exit on exception at 'flash before'

Fix Twister DeviceHandler exit on SerialException when
it connects to the serial device in 'flash before' mode.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
twister: DeviceHandler make DUT counter increment atomic

Twister DeviceHandler - make DUT use counter increment
operation atomic.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
twister: DeviceHandler count test failures on DUTs

Twister DeviceHandler - add test failure counter for how many
test instances have been failed on each DUT (Device Under Test)
when it executes the current test plan.
Output DUT falure counter summary at the end of Twister run.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
twister: PyTest plugin's DUT release sequence change

Change Twister PyTest plugin's test finalizing sequence to release
the DUT it is used as the very last operation, after the Test Instance
status becomes fully updated from the execution results.

This also fix a race condition possible when pytest plugin releases
the DUT and it becomes acquired by another test while the current test
is not yet finalized completely.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
twister: DeviceHandler improve DUT selection

Improve DUT selection at DeviceHandler: for each DUT it counts
how many test instances have been failed on it during the current
twister execution, so the next available DUT will be chosen
ordering the eligible DUTs by less failures occured so far.

The new selection mechanism should increase chances to retry failed
tests on different DUTs, for instance to resolve ploblems when some
DUTs have connectivity or HW issues slowing down test plan execution,
or even block the execution when only one test suite runs whereas
the same first DUT candidate in the list is not working and others
were not chosen.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
dear reviewers, up to your attention

hakehuang commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -484,8 +484,10 @@ def device_is_available(self, instance):
         return None
 
     def make_dut_available(self, dut):
+        if self.instance.status in ["error", "failed"]:
```
same as above

golowanow replied:
to fix this problem twister needs to consider 'failed' test instances at a DUT:

> .. for instance to resolve ploblems when some DUTs have connectivity or HW issues slowing down test plan execution, or even block the execution when only one test suite runs whereas the same first DUT candidate in the list is not working and others were not chosen.

LukaszMrugala commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -473,7 +473,7 @@ def device_is_available(self, instance):
                 avail = True
                 logger.debug(f"Retain DUT:{d.platform}, Id:{d.id}, "
-                             f"counter:{d.counter}")
+                             f"counter:{d.counter}, failures:{d.failures}")
```
Both are necessary, as hardware failure oft shows as test timeouts, which are `failed` statuses, not `error`s.
At the same time, I'd rank them separately; a separate `failures` and `errors` counters, which would be used in sorting, with errors given a higher priority (`sorted()`'s `key` accepts tuples)

golowanow replied:
The proposed change counts both 'error' and 'failed' test instance statuses for each DUT as a failure situation happened there not going deeper into details of what was the origin. It is a single counter to avoid overcomplication, but to stimulate retry attempts on different DUTs choosing less 'suspicious' one as a candidate for the next round. This way it works for any root cause not known at the moment - either at a board, a harness, or a test, or any their combination.

LukaszMrugala reviewed (changes requested):
I'd approve, were it not for the hash.

golowanow commented on scripts/pylib/twister/twisterlib/handlers.py:
```diff
@@ -458,33 +458,40 @@ def monitor_serial(self, ser, halt_event, harness):
-        if not dut_found:
-            raise TwisterException(f"No device to serve as {device} platform.")
-
+        #
```
removed

golowanow replied:
@LukaszMrugala do you have any other objections ?

golowanow commented:
rebased

golowanow commented:
dear reviewers, up to your attention

golowanow commented:
> dear reviewers, up to your attention

just rebased again

golowanow commented:
> > dear reviewers, up to your attention
> 
> just rebased again

and rebased adjusting to #71401

katgiadla reviewed (dismissed):
It works on CI. LGTM

golowanow commented:
> > > dear reviewers, up to your attention
> > 
> > 
> > just rebased again
> 
> and rebased adjusting to #71401

.. and rebased adjusting to #76962
